Userland/Utilities/stty.cpp

```diff
@@ -0,0 +1,616 @@
+        warnln("Save string has an incorrect number of parameters");
+        return 1;
+    }
+    auto parse_hex = [&](const StringView& v) {
+        tcflag_t ret = 0;
+        for (auto c : v) {
+            c = tolower(c);
+            ret *= 16;
+            if (isdigit(c)) {
+                ret += c - '0';
+            } else {
+                VERIFY(c >= 'a' && c <= 'f');
+                ret += c - 'a';
+            }
+        }
+        return ret;
+    };
+
+    t.c_iflag = parse_hex(split[0]);
+    t.c_oflag = parse_hex(split[1]);
+    t.c_cflag = parse_hex(split[2]);
+    t.c_lflag = parse_hex(split[3]);
+    for (size_t i = 0; i < NCCS; ++i) {
+        t.c_cc[i] = (cc_t)parse_hex(split[4 + i]);
+    }
+    t.c_ispeed = parse_hex(split[4 + NCCS]);
+    t.c_ospeed = parse_hex(split[4 + NCCS + 1]);
+    return {};
+}
+
+Result<void, int> apply_modes(size_t parameter_count, char** raw_parameters, termios& t, winsize& w)
+{
+    Vector<StringView> parameters;
+    parameters.ensure_capacity(parameter_count);
+    for (size_t i = 0; i < parameter_count; ++i)
+        parameters.append(StringView(raw_parameters[i]));
+
+    auto parse_baud = [&](size_t idx) -> Optional<speed_t> {
+        auto maybe_numeric_value = parameters[idx].to_uint<uint32_t>();
+        if (maybe_numeric_value.has_value())
+            return numeric_value_to_speed(maybe_numeric_value.value());
+        return {};
+    };
+
+    auto parse_number = [&](size_t idx) -> Optional<cc_t> {
+        return parameters[idx].to_uint<cc_t>();
+    };
+
+    auto looks_like_stty_readable = [&](size_t idx) {
+        bool contains_colon = false;
+        for (auto c : parameters[idx]) {
+            c = tolower(c);
+            if (!isdigit(c) && !(c >= 'a' && c <= 'f') && c != ':')
+                return false;
+            if (c == ':')
+                contains_colon = true;
+        }
+        return contains_colon;
+    };
+
+    auto parse_control_character = [&](size_t idx) -> Result<cc_t, int> {
+        VERIFY(!parameters[idx].is_empty());
+        if (parameters[idx] == "^-" || parameters[idx] == "undef") {
+            // FIXME: disabling characters is a bit wonky right now in TTY.
+            // We should add the _POSIX_VDISABLE macro.
+            return (cc_t)0;
+        } else if (parameters[idx][0] == '^' && parameters[idx].length() == 2) {
+            return (cc_t)toupper(parameters[idx][1]) - 0x40;
+        } else if (parameters[idx].starts_with("0x")) {
+            cc_t value = 0;
+            for (size_t i = 2; i < parameters[idx].length(); ++i) {
+                char ch = tolower(parameters[idx][i]);
+                if (!isdigit(ch) && !(ch >= 'a' && ch <= 'f')) {
+                    warnln("Invalid hexadecimal character code {}", parameters[idx]);
+                    return (int)1;
+                }
+                value = 16 * value + (isdigit(ch)) ? (ch - '0') : (ch - 'a');
+            }
+            return value;
```

Comment:
Should this accept a '0x' prefix when not followed by any hex digits?

Reply:
Fixed, thank you for noticing this.